packages/react-devtools-scheduling-profiler/src/view-base/geometry.js

```diff
@@ -0,0 +1,129 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @flow
+ */
+
+type MutablePoint = {x: number, y: number};
+type MutableSize = {width: number, height: number};
+
+export type Point = $ReadOnly<MutablePoint>;
+export type Size = $ReadOnly<MutableSize>;
+export type Rect = $ReadOnly<{origin: Point, size: Size}>;
+
+/**
+ * Alternative representation of `Rect`.
+ * A tuple of (`top`, `right`, `bottom`, `left`) coordinates.
+ */
+type RectBoundaries = [number, number, number, number];
+
+export const zeroPoint: Point = Object.freeze({x: 0, y: 0});
+export const zeroSize: Size = Object.freeze({width: 0, height: 0});
+export const zeroRect: Rect = Object.freeze({
+  origin: zeroPoint,
+  size: zeroSize,
+});
+
+export function pointEqualToPoint(point1: Point, point2: Point): boolean {
+  return point1.x === point2.x && point1.y === point2.y;
+}
+
+export function sizeEqualToSize(size1: Size, size2: Size): boolean {
+  return size1.width === size2.width && size1.height === size2.height;
+}
+
+export function rectEqualToRect(rect1: Rect, rect2: Rect): boolean {
+  return (
+    pointEqualToPoint(rect1.origin, rect2.origin) &&
+    sizeEqualToSize(rect1.size, rect2.size)
+  );
+}
+
+export function sizeIsValid({width, height}: Size): boolean {
+  return width >= 0 && height >= 0;
+}
+
+export function sizeIsEmpty({width, height}: Size): boolean {
+  return width <= 0 || height <= 0;
+}
+
+function rectToBoundaries(rect: Rect): RectBoundaries {
+  const top = rect.origin.y;
+  const right = rect.origin.x + rect.size.width;
+  const bottom = rect.origin.y + rect.size.height;
+  const left = rect.origin.x;
+  return [top, right, bottom, left];
+}
+
+function boundariesToRect(boundaries: RectBoundaries): Rect {
+  const [top, right, bottom, left] = boundaries;
+  return {
+    origin: {
+      x: left,
+      y: top,
+    },
+    size: {
+      width: right - left,
+      height: bottom - top,
+    },
+  };
+}
+
+export function rectIntersectsRect(rect1: Rect, rect2: Rect): boolean {
+  const [top1, right1, bottom1, left1] = rectToBoundaries(rect1);
+  const [top2, right2, bottom2, left2] = rectToBoundaries(rect2);
+  return !(
+    right1 < left2 ||
+    right2 < left1 ||
+    bottom1 < top2 ||
+    bottom2 < top1
+  );
+}
+
+/**
+ * Prerequisite: rect1 must intersect with rect2.
+ */
+export function rectIntersectionWithRect(rect1: Rect, rect2: Rect): Rect {
```

Comment:
Makes sense! When I wrote this I didn't want to call it `intersectionOfRects` because I thought it'd be more confusing that `unionOfRects` accepts any number of rects whereas this only accepts 2. But I agree that `intersectionOfRects` will be easier to read, especially since it's often called close to `rectIntersectsRect`. I'll rename it